{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3bb64c7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import squidpy as sq\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import decoupler as dc\n",
    "import plotnine as p9\n",
    "import liana as li\n",
    "import numpy as np\n",
    "\n",
    "from pathlib import Path\n",
    "from scipy.sparse import issparse, csr_matrix\n",
    "from liana.method import cellphonedb, natmi\n",
    "from mousipy import translate\n",
    "import datetime"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4e85809",
   "metadata": {},
   "source": [
    "# loading data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "174bd4c9",
   "metadata": {},
   "source": [
    "# Merfish"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "86edb1e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: adata.X seems to be already log-transformed.\n",
      "Index(['Htr7', 'Gzmk', 'Arhgap36', 'Sema3c', 'Rxrg'], dtype='object', name='gene_name')\n",
      "         Method Name Magnitude Score Specificity Score  \\\n",
      "0        CellPhoneDB        lr_means   cellphone_pvals   \n",
      "0         Connectome       expr_prod     scaled_weight   \n",
      "0             log2FC            None          lr_logfc   \n",
      "0              NATMI       expr_prod       spec_weight   \n",
      "0  SingleCellSignalR         lrscore              None   \n",
      "0           CellChat        lr_probs    cellchat_pvals   \n",
      "0     Rank_Aggregate  magnitude_rank  specificity_rank   \n",
      "0     Geometric Mean       lr_gmeans       gmean_pvals   \n",
      "\n",
      "                                           Reference  \n",
      "0  Efremova, M., Vento-Tormo, M., Teichmann, S.A....  \n",
      "0  Raredon, M.S.B., Yang, J., Garritano, J., Wang...  \n",
      "0  Dimitrov, D., Türei, D., Garrido-Rodriguez, M....  \n",
      "0  Hou, R., Denisenko, E., Ong, H.T., Ramilowski,...  \n",
      "0  Cabello-Aguilar, S., Alame, M., Kon-Sun-Tack, ...  \n",
      "0  Jin, S., Guerrero-Juarez, C.F., Zhang, L., Cha...  \n",
      "0  Dimitrov, D., Türei, D., Garrido-Rodriguez, M....  \n",
      "0  CellPhoneDBv2's permutation approach applied t...  \n",
      "['baccin2019', 'cellcall', 'cellchatdb', 'cellinker', 'cellphonedb', 'celltalkdb', 'connectomedb2020', 'consensus', 'embrace', 'guide2pharma', 'hpmr', 'icellnet', 'italk', 'kirouac2010', 'lrdb', 'mouseconsensus', 'ramilowski2015']\n"
     ]
    }
   ],
   "source": [
    "# nemelo by se najit hodne interakci\n",
    "adata = sc.read_h5ad(\"../PRIMUS/data/33_data/transfer/Sofie_bc_prace/data_true/merscope_final_annotated_subseted_gene_symbol.h5ad\")\n",
    "adata\n",
    "\n",
    "adata.obs.columns\n",
    "adata.var\n",
    "adata.layers['raw'] = adata.X.copy()\n",
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "sc.pp.log1p(adata)\n",
    "adata.layers['lognorm'] = adata.X.copy()\n",
    "#just to look how it looks like, especially if we have Ensemble IDs or gene names\n",
    "adata.to_df(layer=\"raw\")\n",
    "\n",
    "adata.var.index = adata.var[\"gene_name\"]\n",
    "\n",
    "#control that the names are now symbols. If not, try adata.var_names = adata.var[\"gene_name\"]\n",
    "\n",
    "adata.to_df(layer=\"raw\")\n",
    "print(adata.var_names[:5])\n",
    "\n",
    "print(li.mt.show_methods())\n",
    "print(li.resource.show_resources())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "bb4ecc88",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1122/1122 [00:02<00:00, 459.55it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found direct orthologs for 1056 genes.\n",
      "Found multiple orthologs for 3 genes.\n",
      "Found no orthologs for 55 genes.\n",
      "Found no index in biomart for 8 genes.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 3/3 [00:08<00:00,  2.87s/it]\n",
      "  0%|          | 0/1 [00:00<?, ?it/s]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:143: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "                                             "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: adata.X seems to be already log-transformed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\r"
     ]
    }
   ],
   "source": [
    "adata.X = adata.layers['raw']\n",
    "humanized_adata = translate(adata,stay_sparse=True)\n",
    "humanized_adata.var\n",
    "\n",
    "humanized_adata.layers['raw'] = humanized_adata.X.copy()\n",
    "sc.pp.normalize_total(humanized_adata, target_sum=1e4)\n",
    "sc.pp.log1p(humanized_adata)\n",
    "humanized_adata.layers['lognorm'] = humanized_adata.X.copy()\n",
    "\n",
    "meta = pd.DataFrame({\n",
    "    'Cell': humanized_adata.obs_names,\n",
    "    'cell_type': humanized_adata.obs['cell_type']\n",
    "})\n",
    "meta.to_csv(\"../output/cpdb/merfish_meta.txt\", sep='\\t', index=False)\n",
    "\n",
    "#something like this will be important for visiumHD because the csv is gonna be HUGE\n",
    "#not tested\n",
    "adata_to_use = humanized_adata.copy()\n",
    "adata_to_use.X = adata_to_use.layers['raw']  # or use .X if raw is not available\n",
    "adata_to_use.var_names_make_unique()\n",
    "\n",
    "# Create the DataFrame\n",
    "counts = pd.DataFrame(\n",
    "    adata_to_use.X.T.toarray() if hasattr(adata_to_use.X, \"toarray\") else adata_to_use.X.T,\n",
    "    index=adata_to_use.var_names,\n",
    "    columns=adata_to_use.obs_names\n",
    ")\n",
    "\n",
    "# Save as HDF5\n",
    "counts.to_hdf(\"../output/cpdb/merfish_counts.h5\", key=\"counts\", mode=\"w\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d6e29151",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organism_ontology_term_id</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>development_stage_ontology_term_id</th>\n",
       "      <th>sex_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>...</th>\n",
       "      <th>is_primary_data</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>assay</th>\n",
       "      <th>disease</th>\n",
       "      <th>organism</th>\n",
       "      <th>sex</th>\n",
       "      <th>tissue</th>\n",
       "      <th>self_reported_ethnicity</th>\n",
       "      <th>development_stage</th>\n",
       "      <th>observation_joinid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1062701771071696533487975728882770248</th>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>C57BL6J-2</td>\n",
       "      <td>MmusDv:0000155</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>na</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0000955</td>\n",
       "      <td>CL:4023051</td>\n",
       "      <td>EFO:0008992</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>vascular leptomeningeal cell</td>\n",
       "      <td>MERFISH</td>\n",
       "      <td>normal</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>male</td>\n",
       "      <td>brain</td>\n",
       "      <td>na</td>\n",
       "      <td>9-week-old stage</td>\n",
       "      <td>G;aLxtIU+O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140347902001698538054266201822142859661</th>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>C57BL6J-2</td>\n",
       "      <td>MmusDv:0000155</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>na</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0000955</td>\n",
       "      <td>CL:4023051</td>\n",
       "      <td>EFO:0008992</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>vascular leptomeningeal cell</td>\n",
       "      <td>MERFISH</td>\n",
       "      <td>normal</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>male</td>\n",
       "      <td>brain</td>\n",
       "      <td>na</td>\n",
       "      <td>9-week-old stage</td>\n",
       "      <td>&amp;5HvbDNa2`</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17912510556127714390369784436317555462</th>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>C57BL6J-2</td>\n",
       "      <td>MmusDv:0000155</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>na</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0000955</td>\n",
       "      <td>CL:0000127</td>\n",
       "      <td>EFO:0008992</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>astrocyte</td>\n",
       "      <td>MERFISH</td>\n",
       "      <td>normal</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>male</td>\n",
       "      <td>brain</td>\n",
       "      <td>na</td>\n",
       "      <td>9-week-old stage</td>\n",
       "      <td>Vhh72ga0fu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202887872709858303957792369114911560476</th>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>C57BL6J-2</td>\n",
       "      <td>MmusDv:0000155</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>na</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0001950</td>\n",
       "      <td>CL:0002453</td>\n",
       "      <td>EFO:0008992</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>oligodendrocyte precursor cell</td>\n",
       "      <td>MERFISH</td>\n",
       "      <td>normal</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>male</td>\n",
       "      <td>neocortex</td>\n",
       "      <td>na</td>\n",
       "      <td>9-week-old stage</td>\n",
       "      <td>9zdojI7}a9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218006636060783985610825029641487829718</th>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>C57BL6J-2</td>\n",
       "      <td>MmusDv:0000155</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>na</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0000955</td>\n",
       "      <td>CL:0000129</td>\n",
       "      <td>EFO:0008992</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>microglial cell</td>\n",
       "      <td>MERFISH</td>\n",
       "      <td>normal</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>male</td>\n",
       "      <td>brain</td>\n",
       "      <td>na</td>\n",
       "      <td>9-week-old stage</td>\n",
       "      <td>b|i|;7j9Us</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5197191772708129939279235494138010912</th>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>C57BL6J-2</td>\n",
       "      <td>MmusDv:0000155</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>na</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0000955</td>\n",
       "      <td>CL:0000617</td>\n",
       "      <td>EFO:0008992</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>GABAergic neuron</td>\n",
       "      <td>MERFISH</td>\n",
       "      <td>normal</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>male</td>\n",
       "      <td>brain</td>\n",
       "      <td>na</td>\n",
       "      <td>9-week-old stage</td>\n",
       "      <td>B%R+ZiqX=?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57045319109046809920494497966207932917</th>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>C57BL6J-2</td>\n",
       "      <td>MmusDv:0000155</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>na</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0000955</td>\n",
       "      <td>CL:0002453</td>\n",
       "      <td>EFO:0008992</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>oligodendrocyte precursor cell</td>\n",
       "      <td>MERFISH</td>\n",
       "      <td>normal</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>male</td>\n",
       "      <td>brain</td>\n",
       "      <td>na</td>\n",
       "      <td>9-week-old stage</td>\n",
       "      <td>DBXdg6XE8v</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77789447341428476484230796188642956517</th>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>C57BL6J-2</td>\n",
       "      <td>MmusDv:0000155</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>na</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0000955</td>\n",
       "      <td>CL:0000617</td>\n",
       "      <td>EFO:0008992</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>GABAergic neuron</td>\n",
       "      <td>MERFISH</td>\n",
       "      <td>normal</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>male</td>\n",
       "      <td>brain</td>\n",
       "      <td>na</td>\n",
       "      <td>9-week-old stage</td>\n",
       "      <td>vs!Aa9qPH!</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95287028017647642274874978777769935669</th>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>C57BL6J-2</td>\n",
       "      <td>MmusDv:0000155</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>na</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0000955</td>\n",
       "      <td>CL:0000115</td>\n",
       "      <td>EFO:0008992</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>endothelial cell</td>\n",
       "      <td>MERFISH</td>\n",
       "      <td>normal</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>male</td>\n",
       "      <td>brain</td>\n",
       "      <td>na</td>\n",
       "      <td>9-week-old stage</td>\n",
       "      <td>F@&gt;Bf{^d$=</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95711223803583725961511638049940645264</th>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>C57BL6J-2</td>\n",
       "      <td>MmusDv:0000155</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>na</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0000955</td>\n",
       "      <td>CL:0002453</td>\n",
       "      <td>EFO:0008992</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>oligodendrocyte precursor cell</td>\n",
       "      <td>MERFISH</td>\n",
       "      <td>normal</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>male</td>\n",
       "      <td>brain</td>\n",
       "      <td>na</td>\n",
       "      <td>9-week-old stage</td>\n",
       "      <td>9fES6SCChQ</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>43496 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        organism_ontology_term_id   donor_id  \\\n",
       "1062701771071696533487975728882770248             NCBITaxon:10090  C57BL6J-2   \n",
       "140347902001698538054266201822142859661           NCBITaxon:10090  C57BL6J-2   \n",
       "17912510556127714390369784436317555462            NCBITaxon:10090  C57BL6J-2   \n",
       "202887872709858303957792369114911560476           NCBITaxon:10090  C57BL6J-2   \n",
       "218006636060783985610825029641487829718           NCBITaxon:10090  C57BL6J-2   \n",
       "...                                                           ...        ...   \n",
       "5197191772708129939279235494138010912             NCBITaxon:10090  C57BL6J-2   \n",
       "57045319109046809920494497966207932917            NCBITaxon:10090  C57BL6J-2   \n",
       "77789447341428476484230796188642956517            NCBITaxon:10090  C57BL6J-2   \n",
       "95287028017647642274874978777769935669            NCBITaxon:10090  C57BL6J-2   \n",
       "95711223803583725961511638049940645264            NCBITaxon:10090  C57BL6J-2   \n",
       "\n",
       "                                        development_stage_ontology_term_id  \\\n",
       "1062701771071696533487975728882770248                       MmusDv:0000155   \n",
       "140347902001698538054266201822142859661                     MmusDv:0000155   \n",
       "17912510556127714390369784436317555462                      MmusDv:0000155   \n",
       "202887872709858303957792369114911560476                     MmusDv:0000155   \n",
       "218006636060783985610825029641487829718                     MmusDv:0000155   \n",
       "...                                                                    ...   \n",
       "5197191772708129939279235494138010912                       MmusDv:0000155   \n",
       "57045319109046809920494497966207932917                      MmusDv:0000155   \n",
       "77789447341428476484230796188642956517                      MmusDv:0000155   \n",
       "95287028017647642274874978777769935669                      MmusDv:0000155   \n",
       "95711223803583725961511638049940645264                      MmusDv:0000155   \n",
       "\n",
       "                                        sex_ontology_term_id  \\\n",
       "1062701771071696533487975728882770248           PATO:0000384   \n",
       "140347902001698538054266201822142859661         PATO:0000384   \n",
       "17912510556127714390369784436317555462          PATO:0000384   \n",
       "202887872709858303957792369114911560476         PATO:0000384   \n",
       "218006636060783985610825029641487829718         PATO:0000384   \n",
       "...                                                      ...   \n",
       "5197191772708129939279235494138010912           PATO:0000384   \n",
       "57045319109046809920494497966207932917          PATO:0000384   \n",
       "77789447341428476484230796188642956517          PATO:0000384   \n",
       "95287028017647642274874978777769935669          PATO:0000384   \n",
       "95711223803583725961511638049940645264          PATO:0000384   \n",
       "\n",
       "                                        self_reported_ethnicity_ontology_term_id  \\\n",
       "1062701771071696533487975728882770248                                         na   \n",
       "140347902001698538054266201822142859661                                       na   \n",
       "17912510556127714390369784436317555462                                        na   \n",
       "202887872709858303957792369114911560476                                       na   \n",
       "218006636060783985610825029641487829718                                       na   \n",
       "...                                                                          ...   \n",
       "5197191772708129939279235494138010912                                         na   \n",
       "57045319109046809920494497966207932917                                        na   \n",
       "77789447341428476484230796188642956517                                        na   \n",
       "95287028017647642274874978777769935669                                        na   \n",
       "95711223803583725961511638049940645264                                        na   \n",
       "\n",
       "                                        disease_ontology_term_id  \\\n",
       "1062701771071696533487975728882770248               PATO:0000461   \n",
       "140347902001698538054266201822142859661             PATO:0000461   \n",
       "17912510556127714390369784436317555462              PATO:0000461   \n",
       "202887872709858303957792369114911560476             PATO:0000461   \n",
       "218006636060783985610825029641487829718             PATO:0000461   \n",
       "...                                                          ...   \n",
       "5197191772708129939279235494138010912               PATO:0000461   \n",
       "57045319109046809920494497966207932917              PATO:0000461   \n",
       "77789447341428476484230796188642956517              PATO:0000461   \n",
       "95287028017647642274874978777769935669              PATO:0000461   \n",
       "95711223803583725961511638049940645264              PATO:0000461   \n",
       "\n",
       "                                        tissue_ontology_term_id  \\\n",
       "1062701771071696533487975728882770248            UBERON:0000955   \n",
       "140347902001698538054266201822142859661          UBERON:0000955   \n",
       "17912510556127714390369784436317555462           UBERON:0000955   \n",
       "202887872709858303957792369114911560476          UBERON:0001950   \n",
       "218006636060783985610825029641487829718          UBERON:0000955   \n",
       "...                                                         ...   \n",
       "5197191772708129939279235494138010912            UBERON:0000955   \n",
       "57045319109046809920494497966207932917           UBERON:0000955   \n",
       "77789447341428476484230796188642956517           UBERON:0000955   \n",
       "95287028017647642274874978777769935669           UBERON:0000955   \n",
       "95711223803583725961511638049940645264           UBERON:0000955   \n",
       "\n",
       "                                        cell_type_ontology_term_id  \\\n",
       "1062701771071696533487975728882770248                   CL:4023051   \n",
       "140347902001698538054266201822142859661                 CL:4023051   \n",
       "17912510556127714390369784436317555462                  CL:0000127   \n",
       "202887872709858303957792369114911560476                 CL:0002453   \n",
       "218006636060783985610825029641487829718                 CL:0000129   \n",
       "...                                                            ...   \n",
       "5197191772708129939279235494138010912                   CL:0000617   \n",
       "57045319109046809920494497966207932917                  CL:0002453   \n",
       "77789447341428476484230796188642956517                  CL:0000617   \n",
       "95287028017647642274874978777769935669                  CL:0000115   \n",
       "95711223803583725961511638049940645264                  CL:0002453   \n",
       "\n",
       "                                        assay_ontology_term_id  \\\n",
       "1062701771071696533487975728882770248              EFO:0008992   \n",
       "140347902001698538054266201822142859661            EFO:0008992   \n",
       "17912510556127714390369784436317555462             EFO:0008992   \n",
       "202887872709858303957792369114911560476            EFO:0008992   \n",
       "218006636060783985610825029641487829718            EFO:0008992   \n",
       "...                                                        ...   \n",
       "5197191772708129939279235494138010912              EFO:0008992   \n",
       "57045319109046809920494497966207932917             EFO:0008992   \n",
       "77789447341428476484230796188642956517             EFO:0008992   \n",
       "95287028017647642274874978777769935669             EFO:0008992   \n",
       "95711223803583725961511638049940645264             EFO:0008992   \n",
       "\n",
       "                                        suspension_type  ... is_primary_data  \\\n",
       "1062701771071696533487975728882770248                na  ...            True   \n",
       "140347902001698538054266201822142859661              na  ...            True   \n",
       "17912510556127714390369784436317555462               na  ...            True   \n",
       "202887872709858303957792369114911560476              na  ...            True   \n",
       "218006636060783985610825029641487829718              na  ...            True   \n",
       "...                                                 ...  ...             ...   \n",
       "5197191772708129939279235494138010912                na  ...            True   \n",
       "57045319109046809920494497966207932917               na  ...            True   \n",
       "77789447341428476484230796188642956517               na  ...            True   \n",
       "95287028017647642274874978777769935669               na  ...            True   \n",
       "95711223803583725961511638049940645264               na  ...            True   \n",
       "\n",
       "                                                              cell_type  \\\n",
       "1062701771071696533487975728882770248      vascular leptomeningeal cell   \n",
       "140347902001698538054266201822142859661    vascular leptomeningeal cell   \n",
       "17912510556127714390369784436317555462                        astrocyte   \n",
       "202887872709858303957792369114911560476  oligodendrocyte precursor cell   \n",
       "218006636060783985610825029641487829718                 microglial cell   \n",
       "...                                                                 ...   \n",
       "5197191772708129939279235494138010912                  GABAergic neuron   \n",
       "57045319109046809920494497966207932917   oligodendrocyte precursor cell   \n",
       "77789447341428476484230796188642956517                 GABAergic neuron   \n",
       "95287028017647642274874978777769935669                 endothelial cell   \n",
       "95711223803583725961511638049940645264   oligodendrocyte precursor cell   \n",
       "\n",
       "                                           assay  disease      organism   sex  \\\n",
       "1062701771071696533487975728882770248    MERFISH   normal  Mus musculus  male   \n",
       "140347902001698538054266201822142859661  MERFISH   normal  Mus musculus  male   \n",
       "17912510556127714390369784436317555462   MERFISH   normal  Mus musculus  male   \n",
       "202887872709858303957792369114911560476  MERFISH   normal  Mus musculus  male   \n",
       "218006636060783985610825029641487829718  MERFISH   normal  Mus musculus  male   \n",
       "...                                          ...      ...           ...   ...   \n",
       "5197191772708129939279235494138010912    MERFISH   normal  Mus musculus  male   \n",
       "57045319109046809920494497966207932917   MERFISH   normal  Mus musculus  male   \n",
       "77789447341428476484230796188642956517   MERFISH   normal  Mus musculus  male   \n",
       "95287028017647642274874978777769935669   MERFISH   normal  Mus musculus  male   \n",
       "95711223803583725961511638049940645264   MERFISH   normal  Mus musculus  male   \n",
       "\n",
       "                                            tissue self_reported_ethnicity  \\\n",
       "1062701771071696533487975728882770248        brain                      na   \n",
       "140347902001698538054266201822142859661      brain                      na   \n",
       "17912510556127714390369784436317555462       brain                      na   \n",
       "202887872709858303957792369114911560476  neocortex                      na   \n",
       "218006636060783985610825029641487829718      brain                      na   \n",
       "...                                            ...                     ...   \n",
       "5197191772708129939279235494138010912        brain                      na   \n",
       "57045319109046809920494497966207932917       brain                      na   \n",
       "77789447341428476484230796188642956517       brain                      na   \n",
       "95287028017647642274874978777769935669       brain                      na   \n",
       "95711223803583725961511638049940645264       brain                      na   \n",
       "\n",
       "                                        development_stage  observation_joinid  \n",
       "1062701771071696533487975728882770248    9-week-old stage          G;aLxtIU+O  \n",
       "140347902001698538054266201822142859661  9-week-old stage          &5HvbDNa2`  \n",
       "17912510556127714390369784436317555462   9-week-old stage          Vhh72ga0fu  \n",
       "202887872709858303957792369114911560476  9-week-old stage          9zdojI7}a9  \n",
       "218006636060783985610825029641487829718  9-week-old stage          b|i|;7j9Us  \n",
       "...                                                   ...                 ...  \n",
       "5197191772708129939279235494138010912    9-week-old stage          B%R+ZiqX=?  \n",
       "57045319109046809920494497966207932917   9-week-old stage          DBXdg6XE8v  \n",
       "77789447341428476484230796188642956517   9-week-old stage          vs!Aa9qPH!  \n",
       "95287028017647642274874978777769935669   9-week-old stage          F@>Bf{^d$=  \n",
       "95711223803583725961511638049940645264   9-week-old stage          9fES6SCChQ  \n",
       "\n",
       "[43496 rows x 29 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "7415bb82",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cellphone START\n",
      "2025-04-28 22:08:02.517574\n",
      "Reading user files...\n",
      "The following user files were loaded successfully:\n",
      "../output/cpdb/merfish_counts.h5\n",
      "../output/cpdb/merfish_meta.txt\n",
      "[ ][CORE][28/04/25-22:08:03][INFO] [Cluster Statistical Analysis] Threshold:0.1 Iterations:1000 Debug-seed:-1 Threads:4 Precision:3\n",
      "[ ][CORE][28/04/25-22:08:03][INFO] Running Real Analysis\n",
      "[ ][CORE][28/04/25-22:08:03][INFO] Running Statistical Analysis\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1000/1000 [03:13<00:00,  5.18it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ ][CORE][28/04/25-22:11:19][INFO] Building Pvalues result\n",
      "[ ][CORE][28/04/25-22:11:19][INFO] Building results\n",
      "Saved deconvoluted to ../output/cpdb/merfish_cpdb/statistical_analysis_deconvoluted_04_28_2025_221119.txt\n",
      "Saved deconvoluted_percents to ../output/cpdb/merfish_cpdb/statistical_analysis_deconvoluted_percents_04_28_2025_221119.txt\n",
      "Saved means to ../output/cpdb/merfish_cpdb/statistical_analysis_means_04_28_2025_221119.txt\n",
      "Saved pvalues to ../output/cpdb/merfish_cpdb/statistical_analysis_pvalues_04_28_2025_221119.txt\n",
      "Saved significant_means to ../output/cpdb/merfish_cpdb/statistical_analysis_significant_means_04_28_2025_221119.txt\n",
      "cellphone END\n",
      "2025-04-28 22:11:19.927620\n"
     ]
    }
   ],
   "source": [
    "from cellphonedb.src.core.methods import cpdb_statistical_analysis_method\n",
    "\n",
    "#all filtered - nenasel, nema tf\n",
    "\n",
    "\n",
    "print(\"cellphone START\")\n",
    "print(datetime.datetime.now())\n",
    "cpdb_results = cpdb_statistical_analysis_method.call(\n",
    "    cpdb_file_path=\"../data/cellphonedb.zip\",  # use default database (or provide path to custom .zip)\n",
    "    meta_file_path=\"../output/cpdb/merfish_meta.txt\",\n",
    "    counts_file_path=\"../output/cpdb/merfish_counts.h5\",\n",
    "    counts_data=\"gene_name\",  # or 'ensembl'\n",
    "    output_path=\"../output/cpdb/merfish_cpdb\")\n",
    "print(\"cellphone END\")\n",
    "print(datetime.datetime.now())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63e0790f",
   "metadata": {},
   "source": [
    "# Xenium"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "97168eb0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 2299 × 5006\n",
       "    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'genomic_control_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'nucleus_count', 'segmentation_method', 'region', 'z_level', 'cell_labels', 'n_genes', 'n_counts', 'cell_groups'\n",
       "    var: 'gene_ids', 'feature_types', 'genome'\n",
       "    uns: 'log1p', 'spatialdata_attrs'\n",
       "    obsm: 'spatial'\n",
       "    layers: 'lognorm', 'raw'"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "##loading data\n",
    "adata = sc.read_h5ad(\"../PRIMUS/data/33_data/transfer/Sofie_bc_prace/data_true/adata_xenium_final_annotated.h5ad\")\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e245d6a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "humanized_adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "fedf746f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 5006/5006 [00:10<00:00, 459.91it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found direct orthologs for 4746 genes.\n",
      "Found multiple orthologs for 76 genes.\n",
      "Found no orthologs for 126 genes.\n",
      "Found no index in biomart for 58 genes.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 29%|██▉       | 22/76 [00:14<00:35,  1.50it/s]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 46%|████▌     | 35/76 [00:24<00:21,  1.91it/s]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 62%|██████▏   | 47/76 [00:31<00:15,  1.90it/s]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 74%|███████▎  | 56/76 [00:35<00:09,  2.09it/s]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 87%|████████▋ | 66/76 [00:44<00:07,  1.27it/s]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "100%|██████████| 76/76 [00:49<00:00,  1.55it/s]\n",
      "  0%|          | 0/18 [00:00<?, ?it/s]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:143: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "                                               \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: adata.X seems to be already log-transformed.\n"
     ]
    }
   ],
   "source": [
    "adata.X = adata.layers['raw']\n",
    "humanized_adata = translate(adata,stay_sparse=True)\n",
    "humanized_adata.var\n",
    "\n",
    "humanized_adata.layers['raw'] = humanized_adata.X.copy()\n",
    "sc.pp.normalize_total(humanized_adata, target_sum=1e4)\n",
    "sc.pp.log1p(humanized_adata)\n",
    "humanized_adata.layers['lognorm'] = humanized_adata.X.copy()\n",
    "\n",
    "meta = pd.DataFrame({\n",
    "    'Cell': humanized_adata.obs_names,\n",
    "    'cell_type': humanized_adata.obs['cell_groups']\n",
    "})\n",
    "meta.to_csv(\"../output/cpdb/xenium_meta.txt\", sep='\\t', index=False)\n",
    "\n",
    "#something like this will be important for visiumHD because the csv is gonna be HUGE\n",
    "#not tested\n",
    "adata_to_use = humanized_adata.copy()\n",
    "adata_to_use.X = adata_to_use.layers['raw']  # or use .X if raw is not available\n",
    "adata_to_use.var_names_make_unique()\n",
    "\n",
    "# Create the DataFrame\n",
    "counts = pd.DataFrame(\n",
    "    adata_to_use.X.T.toarray() if hasattr(adata_to_use.X, \"toarray\") else adata_to_use.X.T,\n",
    "    index=adata_to_use.var_names,\n",
    "    columns=adata_to_use.obs_names\n",
    ")\n",
    "\n",
    "# Save as HDF5\n",
    "counts.to_hdf(\"../output/cpdb/xenium_counts.h5\", key=\"counts\", mode=\"w\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "463f5ead",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cellphone START\n",
      "2025-04-28 22:12:20.964999\n",
      "Reading user files...\n",
      "The following user files were loaded successfully:\n",
      "../output/cpdb/xenium_counts.h5\n",
      "../output/cpdb/xenium_meta.txt\n",
      "[ ][CORE][28/04/25-22:12:21][INFO] [Cluster Statistical Analysis] Threshold:0 Iterations:1000 Debug-seed:-1 Threads:4 Precision:3\n",
      "[ ][CORE][28/04/25-22:12:21][INFO] Running Real Analysis\n",
      "[ ][CORE][28/04/25-22:12:21][INFO] Running Statistical Analysis\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1000/1000 [00:16<00:00, 60.94it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ ][CORE][28/04/25-22:12:40][INFO] Building Pvalues result\n",
      "[ ][CORE][28/04/25-22:12:40][INFO] Building results\n",
      "Saved deconvoluted to ../output/cpdb/xenium_cpdb/statistical_analysis_deconvoluted_04_28_2025_221240.txt\n",
      "Saved deconvoluted_percents to ../output/cpdb/xenium_cpdb/statistical_analysis_deconvoluted_percents_04_28_2025_221240.txt\n",
      "Saved means to ../output/cpdb/xenium_cpdb/statistical_analysis_means_04_28_2025_221240.txt\n",
      "Saved pvalues to ../output/cpdb/xenium_cpdb/statistical_analysis_pvalues_04_28_2025_221240.txt\n",
      "Saved significant_means to ../output/cpdb/xenium_cpdb/statistical_analysis_significant_means_04_28_2025_221240.txt\n",
      "cellphone END\n",
      "2025-04-28 22:12:41.743750\n"
     ]
    }
   ],
   "source": [
    "##nenaslo, prisna metoda\n",
    "\n",
    "from cellphonedb.src.core.methods import cpdb_statistical_analysis_method\n",
    "from cellphonedb.src.core.methods import cpdb_analysis_method\n",
    "\n",
    "\n",
    "print(\"cellphone START\")\n",
    "print(datetime.datetime.now())\n",
    "cpdb_results = cpdb_statistical_analysis_method.call(\n",
    "    cpdb_file_path=\"../data/cellphonedb.zip\",  # use default database (or provide path to custom .zip)\n",
    "    meta_file_path=\"../output/cpdb/xenium_meta.txt\",\n",
    "    counts_file_path=\"../output/cpdb/xenium_counts.h5\",\n",
    "    counts_data=\"hgnc_symbol\",  # or 'ensembl' or gene_name\n",
    "    output_path=\"../output/cpdb/xenium_cpdb\", threshold=0)\n",
    "\n",
    "print(\"cellphone END\")\n",
    "print(datetime.datetime.now())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d382df91",
   "metadata": {},
   "source": [
    "# Cosmx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4bd78ee3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1_1         Endo\n",
       "2_1         VLMC\n",
       "3_1        Astro\n",
       "4_1         Endo\n",
       "5_1         Endo\n",
       "           ...  \n",
       "311_130    Astro\n",
       "312_130    Meis2\n",
       "313_130     VLMC\n",
       "314_130     VLMC\n",
       "315_130     VLMC\n",
       "Name: cell_groups, Length: 48556, dtype: category\n",
       "Categories (23, object): ['Astro', 'CR', 'Endo', 'L2/3 IT', ..., 'Sncg', 'Sst', 'VLMC', 'Vip']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = sc.read_h5ad(\"../PRIMUS/data/33_data/transfer/Sofie_bc_prace/data_true/cosmx_final_annotated.h5ad\")\n",
    "adata\n",
    "adata.obs.cell_groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2b86da49",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 960/960 [00:02<00:00, 456.59it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found direct orthologs for 910 genes.\n",
      "Found multiple orthologs for 10 genes.\n",
      "Found no orthologs for 22 genes.\n",
      "Found no index in biomart for 18 genes.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/10 [00:00<?, ?it/s]/home/blizkovs/.local/lib/python3.12/site-packages/mousipy/mousipy.py:205: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "/home/blizkovs/.local/lib/python3.12/site-packages/mousipy/mousipy.py:205: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      " 10%|█         | 1/10 [00:03<00:30,  3.43s/it]/home/blizkovs/.local/lib/python3.12/site-packages/mousipy/mousipy.py:205: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "/home/blizkovs/.local/lib/python3.12/site-packages/mousipy/mousipy.py:205: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "/home/blizkovs/.local/lib/python3.12/site-packages/mousipy/mousipy.py:205: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "/home/blizkovs/.local/lib/python3.12/site-packages/mousipy/mousipy.py:205: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "/home/blizkovs/.local/lib/python3.12/site-packages/mousipy/mousipy.py:205: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "/home/blizkovs/.local/lib/python3.12/site-packages/mousipy/mousipy.py:205: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      " 20%|██        | 2/10 [00:13<00:59,  7.42s/it]/home/blizkovs/.local/lib/python3.12/site-packages/mousipy/mousipy.py:205: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "/home/blizkovs/.local/lib/python3.12/site-packages/mousipy/mousipy.py:205: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      " 30%|███       | 3/10 [00:17<00:39,  5.60s/it]/home/blizkovs/.local/lib/python3.12/site-packages/mousipy/mousipy.py:205: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "/home/blizkovs/.local/lib/python3.12/site-packages/mousipy/mousipy.py:205: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      " 40%|████      | 4/10 [00:20<00:28,  4.75s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 50%|█████     | 5/10 [00:23<00:19,  3.93s/it]/home/blizkovs/.local/lib/python3.12/site-packages/mousipy/mousipy.py:205: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "/home/blizkovs/.local/lib/python3.12/site-packages/mousipy/mousipy.py:205: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      " 60%|██████    | 6/10 [00:26<00:15,  3.77s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 70%|███████   | 7/10 [00:27<00:08,  2.83s/it]/home/blizkovs/.local/lib/python3.12/site-packages/mousipy/mousipy.py:205: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "/home/blizkovs/.local/lib/python3.12/site-packages/mousipy/mousipy.py:205: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      " 80%|████████  | 8/10 [00:30<00:06,  3.03s/it]/home/blizkovs/.local/lib/python3.12/site-packages/mousipy/mousipy.py:205: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "/home/blizkovs/.local/lib/python3.12/site-packages/mousipy/mousipy.py:205: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      " 90%|█████████ | 9/10 [00:34<00:03,  3.18s/it]/home/blizkovs/.local/lib/python3.12/site-packages/mousipy/mousipy.py:205: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "/home/blizkovs/.local/lib/python3.12/site-packages/mousipy/mousipy.py:205: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "100%|██████████| 10/10 [00:37<00:00,  3.78s/it]\n",
      "/home/blizkovs/.local/lib/python3.12/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "/home/blizkovs/.local/lib/python3.12/site-packages/mousipy/mousipy.py:270: FutureWarning: unique with argument that is not not a Series, Index, ExtensionArray, or np.ndarray is deprecated and will raise in a future version.\n",
      "  0%|          | 0/2 [00:00<?, ?it/s]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:143: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "/usr/local/lib/python3.12/dist-packages/scanpy/preprocessing/_normalization.py:233: UserWarning: Some cells have zero counts\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: adata.X seems to be already log-transformed.\n"
     ]
    }
   ],
   "source": [
    "adata.X = adata.layers['counts']\n",
    "humanized_adata = translate(adata,stay_sparse=True)\n",
    "humanized_adata.var\n",
    "\n",
    "humanized_adata.layers['counts'] = humanized_adata.X.copy()\n",
    "sc.pp.normalize_total(humanized_adata, target_sum=1e4)\n",
    "sc.pp.log1p(humanized_adata)\n",
    "humanized_adata.layers['lognorm'] = humanized_adata.X.copy()\n",
    "\n",
    "meta = pd.DataFrame({\n",
    "    'Cell': humanized_adata.obs_names,\n",
    "    'cell_type': humanized_adata.obs['cell_groups']\n",
    "})\n",
    "meta.to_csv(\"../output/cpdb/cosmx_meta.txt\", sep='\\t', index=False)\n",
    "\n",
    "#something like this will be important for visiumHD because the csv is gonna be HUGE\n",
    "#not tested\n",
    "adata_to_use = humanized_adata.copy() ##zmena\n",
    "adata_to_use.X = adata_to_use.layers['lognorm']  # or use .X if raw is not available\n",
    "adata_to_use.var_names_make_unique()\n",
    "\n",
    "# Create the DataFrame\n",
    "counts = pd.DataFrame(\n",
    "    adata_to_use.X.T.toarray() if hasattr(adata_to_use.X, \"toarray\") else adata_to_use.X.T,\n",
    "    index=adata_to_use.var_names,\n",
    "    columns=adata_to_use.obs_names\n",
    ")\n",
    "\n",
    "# Save as HDF5\n",
    "counts.to_hdf(\"../output/cpdb/cosmx_counts.h5\", key=\"counts\", mode=\"w\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3027bd2e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cellphone START\n",
      "2025-04-28 22:52:10.407955\n",
      "Reading user files...\n",
      "The following user files were loaded successfully:\n",
      "../output/cpdb/cosmx_counts.h5\n",
      "../output/cpdb/cosmx_meta.txt\n",
      "[ ][CORE][28/04/25-22:52:11][INFO] [Cluster Statistical Analysis] Threshold:0 Iterations:1000 Debug-seed:-1 Threads:4 Precision:3\n",
      "[ ][CORE][28/04/25-22:52:14][INFO] Running Real Analysis\n",
      "[ ][CORE][28/04/25-22:52:14][INFO] Running Statistical Analysis\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1000/1000 [03:46<00:00,  4.41it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ ][CORE][28/04/25-22:56:01][INFO] Building Pvalues result\n",
      "[ ][CORE][28/04/25-22:56:01][INFO] Building results\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved deconvoluted to ../output/cpdb/cosmx_cpdb/statistical_analysis_deconvoluted_04_28_2025_225601.txt\n",
      "Saved deconvoluted_percents to ../output/cpdb/cosmx_cpdb/statistical_analysis_deconvoluted_percents_04_28_2025_225601.txt\n",
      "Saved means to ../output/cpdb/cosmx_cpdb/statistical_analysis_means_04_28_2025_225601.txt\n",
      "Saved pvalues to ../output/cpdb/cosmx_cpdb/statistical_analysis_pvalues_04_28_2025_225601.txt\n",
      "Saved significant_means to ../output/cpdb/cosmx_cpdb/statistical_analysis_significant_means_04_28_2025_225601.txt\n",
      "cellphone END\n",
      "2025-04-28 22:56:01.656887\n"
     ]
    }
   ],
   "source": [
    "from cellphonedb.src.core.methods import cpdb_statistical_analysis_method\n",
    "from cellphonedb.src.core.methods import cpdb_analysis_method\n",
    "\n",
    "\n",
    "print(\"cellphone START\")\n",
    "print(datetime.datetime.now())\n",
    "cpdb_results = cpdb_statistical_analysis_method.call( \n",
    "    cpdb_file_path=\"../data/cellphonedb.zip\",  # use default database (or provide path to custom .zip)\n",
    "    meta_file_path=\"../output/cpdb/cosmx_meta.txt\",\n",
    "    counts_file_path=\"../output/cpdb/cosmx_counts.h5\",\n",
    "    counts_data=\"hgnc_symbol\",  # or 'ensembl' or gene_name\n",
    "    output_path=\"../output/cpdb/cosmx_cpdb\", threshold=0)\n",
    "\n",
    "print(\"cellphone END\")\n",
    "print(datetime.datetime.now())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7866b38",
   "metadata": {},
   "source": [
    "# Visium"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f2aeb3fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 273789 × 16797\n",
       "    obs: 'in_tissue', 'array_row', 'array_col', 'location_id', 'region', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'n_counts', 'barcode', 'label'\n",
       "    var: 'gene_ids', 'feature_types', 'genome', 'mt', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells'\n",
       "    uns: 'label_colors', 'log1p', 'spatialdata_attrs'\n",
       "    obsm: 'spatial'\n",
       "    layers: 'lognorm'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###ma ensembl\n",
    "adata = sc.read_h5ad(\"../PRIMUS/data/33_data/transfer/Sofie_bc_prace/visium_hd_annotated.h5ad\")\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "bb7092be",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>genome</th>\n",
       "      <th>mt</th>\n",
       "      <th>n_cells_by_counts</th>\n",
       "      <th>mean_counts</th>\n",
       "      <th>log1p_mean_counts</th>\n",
       "      <th>pct_dropout_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>log1p_total_counts</th>\n",
       "      <th>n_cells</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Xkr4</th>\n",
       "      <td>ENSMUSG00000051951</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "      <td>False</td>\n",
       "      <td>5089</td>\n",
       "      <td>0.013351</td>\n",
       "      <td>0.013262</td>\n",
       "      <td>98.706876</td>\n",
       "      <td>5254.0</td>\n",
       "      <td>8.566936</td>\n",
       "      <td>4456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rp1</th>\n",
       "      <td>ENSMUSG00000025900</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "      <td>False</td>\n",
       "      <td>14</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>99.996443</td>\n",
       "      <td>14.0</td>\n",
       "      <td>2.708050</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sox17</th>\n",
       "      <td>ENSMUSG00000025902</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "      <td>False</td>\n",
       "      <td>745</td>\n",
       "      <td>0.001972</td>\n",
       "      <td>0.001970</td>\n",
       "      <td>99.810694</td>\n",
       "      <td>776.0</td>\n",
       "      <td>6.655440</td>\n",
       "      <td>627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lypla1</th>\n",
       "      <td>ENSMUSG00000025903</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "      <td>False</td>\n",
       "      <td>2108</td>\n",
       "      <td>0.005425</td>\n",
       "      <td>0.005410</td>\n",
       "      <td>99.464353</td>\n",
       "      <td>2135.0</td>\n",
       "      <td>7.666690</td>\n",
       "      <td>1844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tcea1</th>\n",
       "      <td>ENSMUSG00000033813</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "      <td>False</td>\n",
       "      <td>12070</td>\n",
       "      <td>0.031943</td>\n",
       "      <td>0.031444</td>\n",
       "      <td>96.932991</td>\n",
       "      <td>12571.0</td>\n",
       "      <td>9.439227</td>\n",
       "      <td>10403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Nd4</th>\n",
       "      <td>ENSMUSG00000064363</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "      <td>False</td>\n",
       "      <td>130537</td>\n",
       "      <td>0.439360</td>\n",
       "      <td>0.364198</td>\n",
       "      <td>66.830308</td>\n",
       "      <td>172907.0</td>\n",
       "      <td>12.060514</td>\n",
       "      <td>102133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Nd5</th>\n",
       "      <td>ENSMUSG00000064367</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "      <td>False</td>\n",
       "      <td>137032</td>\n",
       "      <td>0.474904</td>\n",
       "      <td>0.388593</td>\n",
       "      <td>65.179917</td>\n",
       "      <td>186895.0</td>\n",
       "      <td>12.138308</td>\n",
       "      <td>108835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Nd6</th>\n",
       "      <td>ENSMUSG00000064368</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "      <td>False</td>\n",
       "      <td>22495</td>\n",
       "      <td>0.060016</td>\n",
       "      <td>0.058284</td>\n",
       "      <td>94.283979</td>\n",
       "      <td>23619.0</td>\n",
       "      <td>10.069849</td>\n",
       "      <td>17290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Cytb</th>\n",
       "      <td>ENSMUSG00000064370</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "      <td>False</td>\n",
       "      <td>221308</td>\n",
       "      <td>0.964616</td>\n",
       "      <td>0.675297</td>\n",
       "      <td>43.765230</td>\n",
       "      <td>379618.0</td>\n",
       "      <td>12.846924</td>\n",
       "      <td>171615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Vamp7</th>\n",
       "      <td>ENSMUSG00000051412</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "      <td>False</td>\n",
       "      <td>4131</td>\n",
       "      <td>0.010741</td>\n",
       "      <td>0.010684</td>\n",
       "      <td>98.950305</td>\n",
       "      <td>4227.0</td>\n",
       "      <td>8.349484</td>\n",
       "      <td>3571</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16797 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   gene_ids    feature_types genome     mt  n_cells_by_counts  \\\n",
       "Xkr4     ENSMUSG00000051951  Gene Expression   mm10  False               5089   \n",
       "Rp1      ENSMUSG00000025900  Gene Expression   mm10  False                 14   \n",
       "Sox17    ENSMUSG00000025902  Gene Expression   mm10  False                745   \n",
       "Lypla1   ENSMUSG00000025903  Gene Expression   mm10  False               2108   \n",
       "Tcea1    ENSMUSG00000033813  Gene Expression   mm10  False              12070   \n",
       "...                     ...              ...    ...    ...                ...   \n",
       "mt-Nd4   ENSMUSG00000064363  Gene Expression   mm10  False             130537   \n",
       "mt-Nd5   ENSMUSG00000064367  Gene Expression   mm10  False             137032   \n",
       "mt-Nd6   ENSMUSG00000064368  Gene Expression   mm10  False              22495   \n",
       "mt-Cytb  ENSMUSG00000064370  Gene Expression   mm10  False             221308   \n",
       "Vamp7    ENSMUSG00000051412  Gene Expression   mm10  False               4131   \n",
       "\n",
       "         mean_counts  log1p_mean_counts  pct_dropout_by_counts  total_counts  \\\n",
       "Xkr4        0.013351           0.013262              98.706876        5254.0   \n",
       "Rp1         0.000036           0.000036              99.996443          14.0   \n",
       "Sox17       0.001972           0.001970              99.810694         776.0   \n",
       "Lypla1      0.005425           0.005410              99.464353        2135.0   \n",
       "Tcea1       0.031943           0.031444              96.932991       12571.0   \n",
       "...              ...                ...                    ...           ...   \n",
       "mt-Nd4      0.439360           0.364198              66.830308      172907.0   \n",
       "mt-Nd5      0.474904           0.388593              65.179917      186895.0   \n",
       "mt-Nd6      0.060016           0.058284              94.283979       23619.0   \n",
       "mt-Cytb     0.964616           0.675297              43.765230      379618.0   \n",
       "Vamp7       0.010741           0.010684              98.950305        4227.0   \n",
       "\n",
       "         log1p_total_counts  n_cells  \n",
       "Xkr4               8.566936     4456  \n",
       "Rp1                2.708050       12  \n",
       "Sox17              6.655440      627  \n",
       "Lypla1             7.666690     1844  \n",
       "Tcea1              9.439227    10403  \n",
       "...                     ...      ...  \n",
       "mt-Nd4            12.060514   102133  \n",
       "mt-Nd5            12.138308   108835  \n",
       "mt-Nd6            10.069849    17290  \n",
       "mt-Cytb           12.846924   171615  \n",
       "Vamp7              8.349484     3571  \n",
       "\n",
       "[16797 rows x 11 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "293444d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "preparing START\n",
      "2025-04-30 10:31:41.294555\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 22%|██▏       | 3744/16797 [02:38<8:19:45,  2.30s/it]IOStream.flush timed out\n",
      "100%|██████████| 16797/16797 [06:39<00:00, 41.99it/s]   \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found direct orthologs for 14789 genes.\n",
      "Found multiple orthologs for 303 genes.\n",
      "Found no orthologs for 846 genes.\n",
      "Found no index in biomart for 859 genes.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 1/303 [02:48<14:07:02, 168.29s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  2%|▏         | 5/303 [14:01<17:12:57, 207.98s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  4%|▍         | 12/303 [36:29<16:41:28, 206.49s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  5%|▍         | 14/303 [39:21<12:11:47, 151.93s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 10%|▉         | 29/303 [1:32:33<15:47:34, 207.50s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 12%|█▏        | 36/303 [1:50:54<13:16:09, 178.91s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 13%|█▎        | 38/303 [2:09:07<30:34:20, 415.32s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 15%|█▍        | 45/303 [2:24:58<13:23:51, 186.95s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 15%|█▌        | 46/303 [2:55:23<48:25:45, 678.39s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 20%|██        | 62/303 [4:05:28<21:56:25, 327.74s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 22%|██▏       | 68/303 [4:21:04<12:18:17, 188.50s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 26%|██▌       | 79/303 [4:40:45<12:29:45, 200.83s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 27%|██▋       | 81/303 [4:50:39<17:01:46, 276.15s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 28%|██▊       | 84/303 [5:01:59<17:10:28, 282.32s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 29%|██▉       | 89/303 [6:01:14<42:01:08, 706.86s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 32%|███▏      | 96/303 [6:18:57<13:26:36, 233.80s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 33%|███▎      | 100/303 [6:31:50<11:57:16, 212.00s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 34%|███▎      | 102/303 [6:37:35<11:34:02, 207.18s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 39%|███▉      | 119/303 [7:08:02<9:49:25, 192.20s/it] /usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 41%|████      | 123/303 [7:35:07<20:06:26, 402.15s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 41%|████▏     | 125/303 [7:39:26<13:35:04, 274.75s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 43%|████▎     | 129/303 [7:45:17<7:27:36, 154.35s/it] /usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 44%|████▎     | 132/303 [7:51:03<6:41:29, 140.87s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 45%|████▍     | 135/303 [8:04:01<13:07:21, 281.20s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 48%|████▊     | 146/303 [8:33:11<8:08:51, 186.83s/it] /usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 50%|████▉     | 151/303 [8:47:38<7:55:50, 187.83s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 50%|█████     | 152/303 [8:47:44<5:35:29, 133.31s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 52%|█████▏    | 157/303 [9:03:44<7:02:54, 173.80s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 52%|█████▏    | 159/303 [9:06:40<5:30:24, 137.67s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 55%|█████▌    | 167/303 [9:31:15<7:37:59, 202.05s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 56%|█████▌    | 169/303 [9:34:10<5:37:34, 151.16s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 57%|█████▋    | 173/303 [9:37:09<3:12:44, 88.95s/it] /usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 59%|█████▊    | 178/303 [9:50:00<5:20:32, 153.86s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 60%|█████▉    | 181/303 [9:54:20<4:24:44, 130.20s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 62%|██████▏   | 189/303 [10:21:27<7:11:59, 227.37s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 64%|██████▍   | 194/303 [10:28:39<4:34:28, 151.09s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 66%|██████▌   | 199/303 [10:45:51<8:13:10, 284.52s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 67%|██████▋   | 204/303 [10:58:53<7:42:52, 280.53s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 69%|██████▉   | 210/303 [11:02:06<2:16:11, 87.86s/it] /usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 74%|███████▍  | 224/303 [11:30:59<3:41:00, 167.86s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 77%|███████▋  | 232/303 [11:52:37<4:23:44, 222.88s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 80%|███████▉  | 242/303 [12:10:17<3:12:19, 189.17s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 83%|████████▎ | 250/303 [12:37:36<3:17:42, 223.82s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 85%|████████▌ | 259/303 [12:54:58<1:58:27, 161.55s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 87%|████████▋ | 263/303 [13:03:40<1:41:52, 152.80s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 88%|████████▊ | 268/303 [13:15:13<1:31:35, 157.01s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 89%|████████▉ | 271/303 [13:18:09<56:36, 106.13s/it]  /usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 91%|█████████▏| 277/303 [13:35:36<2:02:32, 282.79s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 92%|█████████▏| 279/303 [13:38:39<1:17:01, 192.57s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 93%|█████████▎| 281/303 [13:41:35<53:55, 147.07s/it]  /usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 97%|█████████▋| 294/303 [14:50:59<43:28, 289.79s/it]  /usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 98%|█████████▊| 297/303 [14:56:50<18:51, 188.53s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      " 99%|█████████▉| 300/303 [15:07:03<12:22, 247.59s/it]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "100%|██████████| 303/303 [15:10:03<00:00, 180.21s/it]\n",
      "  0%|          | 0/136 [00:00<?, ?it/s]/usr/lib/python3/dist-packages/scipy/sparse/_index.py:143: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "                                                 \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: adata.X seems to be already log-transformed.\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "The file '../output/cpdb/visium_counts.h5' is already opened.  Please close it before reopening in write mode.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[23], line 32\u001b[0m\n\u001b[1;32m     25\u001b[0m counts \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame(\n\u001b[1;32m     26\u001b[0m     adata_to_use\u001b[38;5;241m.\u001b[39mX\u001b[38;5;241m.\u001b[39mT\u001b[38;5;241m.\u001b[39mtoarray() \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(adata_to_use\u001b[38;5;241m.\u001b[39mX, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtoarray\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;28;01melse\u001b[39;00m adata_to_use\u001b[38;5;241m.\u001b[39mX\u001b[38;5;241m.\u001b[39mT,\n\u001b[1;32m     27\u001b[0m     index\u001b[38;5;241m=\u001b[39madata_to_use\u001b[38;5;241m.\u001b[39mvar_names,\n\u001b[1;32m     28\u001b[0m     columns\u001b[38;5;241m=\u001b[39madata_to_use\u001b[38;5;241m.\u001b[39mobs_names\n\u001b[1;32m     29\u001b[0m )\n\u001b[1;32m     31\u001b[0m \u001b[38;5;66;03m# Save as HDF5\u001b[39;00m\n\u001b[0;32m---> 32\u001b[0m \u001b[43mcounts\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto_hdf\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m../output/cpdb/visium_counts.h5\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkey\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcounts\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmode\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mw\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m     34\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpreparing END\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     35\u001b[0m \u001b[38;5;28mprint\u001b[39m(datetime\u001b[38;5;241m.\u001b[39mdatetime\u001b[38;5;241m.\u001b[39mnow())\n",
      "File \u001b[0;32m~/.local/lib/python3.12/site-packages/pandas/util/_decorators.py:333\u001b[0m, in \u001b[0;36mdeprecate_nonkeyword_arguments.<locals>.decorate.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    327\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(args) \u001b[38;5;241m>\u001b[39m num_allow_args:\n\u001b[1;32m    328\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[1;32m    329\u001b[0m         msg\u001b[38;5;241m.\u001b[39mformat(arguments\u001b[38;5;241m=\u001b[39m_format_argument_list(allow_args)),\n\u001b[1;32m    330\u001b[0m         \u001b[38;5;167;01mFutureWarning\u001b[39;00m,\n\u001b[1;32m    331\u001b[0m         stacklevel\u001b[38;5;241m=\u001b[39mfind_stack_level(),\n\u001b[1;32m    332\u001b[0m     )\n\u001b[0;32m--> 333\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.12/site-packages/pandas/core/generic.py:2855\u001b[0m, in \u001b[0;36mNDFrame.to_hdf\u001b[0;34m(self, path_or_buf, key, mode, complevel, complib, append, format, index, min_itemsize, nan_rep, dropna, data_columns, errors, encoding)\u001b[0m\n\u001b[1;32m   2851\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mpandas\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mio\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m pytables\n\u001b[1;32m   2853\u001b[0m \u001b[38;5;66;03m# Argument 3 to \"to_hdf\" has incompatible type \"NDFrame\"; expected\u001b[39;00m\n\u001b[1;32m   2854\u001b[0m \u001b[38;5;66;03m# \"Union[DataFrame, Series]\" [arg-type]\u001b[39;00m\n\u001b[0;32m-> 2855\u001b[0m \u001b[43mpytables\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto_hdf\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   2856\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpath_or_buf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2857\u001b[0m \u001b[43m    \u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2858\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m  \u001b[49m\u001b[38;5;66;43;03m# type: ignore[arg-type]\u001b[39;49;00m\n\u001b[1;32m   2859\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2860\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcomplevel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcomplevel\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2861\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcomplib\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcomplib\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2862\u001b[0m \u001b[43m    \u001b[49m\u001b[43mappend\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mappend\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2863\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43mformat\u001b[39;49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mformat\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2864\u001b[0m \u001b[43m    \u001b[49m\u001b[43mindex\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mindex\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2865\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmin_itemsize\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmin_itemsize\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2866\u001b[0m \u001b[43m    \u001b[49m\u001b[43mnan_rep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mnan_rep\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2867\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdropna\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdropna\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2868\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdata_columns\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdata_columns\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2869\u001b[0m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2870\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2871\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.12/site-packages/pandas/io/pytables.py:308\u001b[0m, in \u001b[0;36mto_hdf\u001b[0;34m(path_or_buf, key, value, mode, complevel, complib, append, format, index, min_itemsize, nan_rep, dropna, data_columns, errors, encoding)\u001b[0m\n\u001b[1;32m    306\u001b[0m path_or_buf \u001b[38;5;241m=\u001b[39m stringify_path(path_or_buf)\n\u001b[1;32m    307\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(path_or_buf, \u001b[38;5;28mstr\u001b[39m):\n\u001b[0;32m--> 308\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m \u001b[43mHDFStore\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    309\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpath_or_buf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmode\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcomplevel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcomplevel\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcomplib\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcomplib\u001b[49m\n\u001b[1;32m    310\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mas\u001b[39;00m store:\n\u001b[1;32m    311\u001b[0m         f(store)\n\u001b[1;32m    312\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/.local/lib/python3.12/site-packages/pandas/io/pytables.py:585\u001b[0m, in \u001b[0;36mHDFStore.__init__\u001b[0;34m(self, path, mode, complevel, complib, fletcher32, **kwargs)\u001b[0m\n\u001b[1;32m    583\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_fletcher32 \u001b[38;5;241m=\u001b[39m fletcher32\n\u001b[1;32m    584\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_filters \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m--> 585\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mopen\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmode\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.12/site-packages/pandas/io/pytables.py:745\u001b[0m, in \u001b[0;36mHDFStore.open\u001b[0;34m(self, mode, **kwargs)\u001b[0m\n\u001b[1;32m    739\u001b[0m     msg \u001b[38;5;241m=\u001b[39m (\n\u001b[1;32m    740\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCannot open HDF5 file, which is already opened, \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    741\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124meven in read-only mode.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    742\u001b[0m     )\n\u001b[1;32m    743\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(msg)\n\u001b[0;32m--> 745\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_handle \u001b[38;5;241m=\u001b[39m \u001b[43mtables\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mopen_file\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_mode\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/lib/python3/dist-packages/tables/file.py:289\u001b[0m, in \u001b[0;36mopen_file\u001b[0;34m(filename, mode, title, root_uep, filters, **kwargs)\u001b[0m\n\u001b[1;32m    287\u001b[0m         \u001b[38;5;66;03m# 'w' means that we want to destroy existing contents\u001b[39;00m\n\u001b[1;32m    288\u001b[0m         \u001b[38;5;28;01melif\u001b[39;00m mode \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mw\u001b[39m\u001b[38;5;124m'\u001b[39m:\n\u001b[0;32m--> 289\u001b[0m             \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m    290\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mThe file \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m is already opened.  Please \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    291\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mclose it before reopening in write mode.\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m%\u001b[39m filename)\n\u001b[1;32m    293\u001b[0m \u001b[38;5;66;03m# Finally, create the File instance, and return it\u001b[39;00m\n\u001b[1;32m    294\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m File(filename, mode, title, root_uep, filters, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "\u001b[0;31mValueError\u001b[0m: The file '../output/cpdb/visium_counts.h5' is already opened.  Please close it before reopening in write mode."
     ]
    }
   ],
   "source": [
    "print(\"preparing START\")\n",
    "print(datetime.datetime.now())\n",
    "adata.X = adata.layers['lognorm']\n",
    "humanized_adata = translate(adata,stay_sparse=True)\n",
    "humanized_adata.var\n",
    "\n",
    "humanized_adata.layers['lognorm'] = humanized_adata.X.copy()\n",
    "sc.pp.normalize_total(humanized_adata, target_sum=1e4)\n",
    "sc.pp.log1p(humanized_adata)\n",
    "humanized_adata.layers['lognorm'] = humanized_adata.X.copy()\n",
    "\n",
    "meta = pd.DataFrame({\n",
    "    'Cell': humanized_adata.obs_names,\n",
    "    'cell_type': humanized_adata.obs['label']\n",
    "})\n",
    "meta.to_csv(\"../output/cpdb/visium_meta.txt\", sep='\\t', index=False)\n",
    "\n",
    "#something like this will be important for visiumHD because the csv is gonna be HUGE\n",
    "#not tested\n",
    "adata_to_use = humanized_adata.copy()\n",
    "adata_to_use.X = adata_to_use.layers['lognorm']  # or use .X if raw is not available\n",
    "adata_to_use.var_names_make_unique()\n",
    "\n",
    "# Create the DataFrame\n",
    "counts = pd.DataFrame(\n",
    "    adata_to_use.X.T.toarray() if hasattr(adata_to_use.X, \"toarray\") else adata_to_use.X.T,\n",
    "    index=adata_to_use.var_names,\n",
    "    columns=adata_to_use.obs_names\n",
    ")\n",
    "\n",
    "# Save as HDF5\n",
    "counts.to_hdf(\"../output/cpdb/visium_counts.h5\", key=\"counts\", mode=\"w\")\n",
    "\n",
    "print(\"preparing END\")\n",
    "print(datetime.datetime.now())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "9570c761",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cellphone START\n",
      "2025-04-30 09:18:32.144457\n",
      "Reading user files...\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "Attribute 'block0_items_variety' does not exist in node: '/counts'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[18], line 8\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcellphone START\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      6\u001b[0m \u001b[38;5;28mprint\u001b[39m(datetime\u001b[38;5;241m.\u001b[39mdatetime\u001b[38;5;241m.\u001b[39mnow())\n\u001b[0;32m----> 8\u001b[0m cpdb_results \u001b[38;5;241m=\u001b[39m \u001b[43mcpdb_statistical_analysis_method\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcall\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m      9\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcpdb_file_path\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m../data/cellphonedb.zip\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m  \u001b[49m\u001b[38;5;66;43;03m# use default database (or provide path to custom .zip)\u001b[39;49;00m\n\u001b[1;32m     10\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmeta_file_path\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m../output/cpdb/visium_meta.txt\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     11\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcounts_file_path\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m../output/cpdb/visium_counts.h5\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     12\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcounts_data\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mensembl\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m  \u001b[49m\u001b[38;5;66;43;03m# or 'ensembl' or gene_name\u001b[39;49;00m\n\u001b[1;32m     13\u001b[0m \u001b[43m    \u001b[49m\u001b[43moutput_path\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m../output/cpdb/visium_cpdb\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m     16\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcellphone END\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     17\u001b[0m \u001b[38;5;28mprint\u001b[39m(datetime\u001b[38;5;241m.\u001b[39mdatetime\u001b[38;5;241m.\u001b[39mnow())\n",
      "File \u001b[0;32m~/.local/lib/python3.12/site-packages/cellphonedb/src/core/methods/cpdb_statistical_analysis_method.py:105\u001b[0m, in \u001b[0;36mcall\u001b[0;34m(cpdb_file_path, meta_file_path, counts_file_path, counts_data, output_path, microenvs_file_path, active_tfs_file_path, iterations, threshold, threads, debug_seed, result_precision, pvalue, subsampling, subsampling_log, subsampling_num_pc, subsampling_num_cells, separator, debug, output_suffix, score_interactions)\u001b[0m\n\u001b[1;32m    101\u001b[0m interactions, genes, complex_composition, complex_expanded, gene_synonym2gene_name, receptor2tfs \u001b[38;5;241m=\u001b[39m \\\n\u001b[1;32m    102\u001b[0m     db_utils\u001b[38;5;241m.\u001b[39mget_interactions_genes_complex(cpdb_file_path)\n\u001b[1;32m    104\u001b[0m \u001b[38;5;66;03m# Load user files into memory\u001b[39;00m\n\u001b[0;32m--> 105\u001b[0m counts, meta, microenvs, degs, active_tf2cell_types \u001b[38;5;241m=\u001b[39m \u001b[43mfile_utils\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_user_files\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    106\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcounts\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcounts_file_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmeta_fp\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmeta_file_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmicroenvs_fp\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmicroenvs_file_path\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    107\u001b[0m \u001b[43m    \u001b[49m\u001b[43mactive_tfs_fp\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mactive_tfs_file_path\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    108\u001b[0m \u001b[43m    \u001b[49m\u001b[43mgene_synonym2gene_name\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgene_synonym2gene_name\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcounts_data\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcounts_data\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    110\u001b[0m \u001b[38;5;66;03m# add multidata id and means to counts\u001b[39;00m\n\u001b[1;32m    111\u001b[0m counts, counts_relations \u001b[38;5;241m=\u001b[39m cpdb_statistical_analysis_helper\u001b[38;5;241m.\u001b[39madd_multidata_and_means_to_counts(\n\u001b[1;32m    112\u001b[0m     counts, genes, counts_data)\n",
      "File \u001b[0;32m~/.local/lib/python3.12/site-packages/cellphonedb/utils/file_utils.py:434\u001b[0m, in \u001b[0;36mget_user_files\u001b[0;34m(counts, meta_fp, microenvs_fp, degs_fp, active_tfs_fp, gene_synonym2gene_name, counts_data)\u001b[0m\n\u001b[1;32m    432\u001b[0m \u001b[38;5;66;03m# Read user files\u001b[39;00m\n\u001b[1;32m    433\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mReading user files...\u001b[39m\u001b[38;5;124m\"\u001b[39m, flush\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[0;32m--> 434\u001b[0m counts, counts_label \u001b[38;5;241m=\u001b[39m \u001b[43mread_counts\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcounts\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    436\u001b[0m \u001b[38;5;66;03m# N.B. The functionality below has been switched off for the time being, on Kevin's request\u001b[39;00m\n\u001b[1;32m    437\u001b[0m \u001b[38;5;66;03m# In counts df, replace any gene synonyms not in gene_input.csv to gene names that are in gene_input.\u001b[39;00m\n\u001b[1;32m    438\u001b[0m \u001b[38;5;66;03m# if counts_data == \"hgnc_symbol\" or counts_data == \"gene_name\":\u001b[39;00m\n\u001b[1;32m    439\u001b[0m \u001b[38;5;66;03m#     counts.rename(index=gene_synonym2gene_name, inplace=True)\u001b[39;00m\n\u001b[1;32m    441\u001b[0m loaded_user_files\u001b[38;5;241m.\u001b[39mappend(counts_label)\n",
      "File \u001b[0;32m~/.local/lib/python3.12/site-packages/cellphonedb/utils/file_utils.py:395\u001b[0m, in \u001b[0;36mread_counts\u001b[0;34m(obj)\u001b[0m\n\u001b[1;32m    393\u001b[0m     counts_label \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcounts from AnnData object\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    394\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 395\u001b[0m     counts \u001b[38;5;241m=\u001b[39m \u001b[43mread_data_table_from_file\u001b[49m\u001b[43m(\u001b[49m\u001b[43mobj\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindex_column_first\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[1;32m    396\u001b[0m     counts_label \u001b[38;5;241m=\u001b[39m obj\n\u001b[1;32m    397\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m counts, counts_label\n",
      "File \u001b[0;32m~/.local/lib/python3.12/site-packages/cellphonedb/utils/file_utils.py:30\u001b[0m, in \u001b[0;36mread_data_table_from_file\u001b[0;34m(file, index_column_first, separator, dtype, na_values, compression, optional)\u001b[0m\n\u001b[1;32m     28\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m _read_h5ad(file)\n\u001b[1;32m     29\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m file_extension \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m.h5\u001b[39m\u001b[38;5;124m'\u001b[39m:\n\u001b[0;32m---> 30\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read_h5\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfile\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     31\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m file_extension \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m.pickle\u001b[39m\u001b[38;5;124m'\u001b[39m:\n\u001b[1;32m     32\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m _read_pickle(file)\n",
      "File \u001b[0;32m~/.local/lib/python3.12/site-packages/cellphonedb/utils/file_utils.py:101\u001b[0m, in \u001b[0;36m_read_h5\u001b[0;34m(path)\u001b[0m\n\u001b[1;32m    100\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_read_h5\u001b[39m(path: \u001b[38;5;28mstr\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame:\n\u001b[0;32m--> 101\u001b[0m     df \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_hdf\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpath\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    102\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m df\n",
      "File \u001b[0;32m~/.local/lib/python3.12/site-packages/pandas/io/pytables.py:452\u001b[0m, in \u001b[0;36mread_hdf\u001b[0;34m(path_or_buf, key, mode, errors, where, start, stop, columns, iterator, chunksize, **kwargs)\u001b[0m\n\u001b[1;32m    447\u001b[0m                 \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m    448\u001b[0m                     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mkey must be provided when HDF5 \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    449\u001b[0m                     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfile contains multiple datasets.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    450\u001b[0m                 )\n\u001b[1;32m    451\u001b[0m         key \u001b[38;5;241m=\u001b[39m candidate_only_group\u001b[38;5;241m.\u001b[39m_v_pathname\n\u001b[0;32m--> 452\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mstore\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mselect\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    453\u001b[0m \u001b[43m        \u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    454\u001b[0m \u001b[43m        \u001b[49m\u001b[43mwhere\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mwhere\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    455\u001b[0m \u001b[43m        \u001b[49m\u001b[43mstart\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstart\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    456\u001b[0m \u001b[43m        \u001b[49m\u001b[43mstop\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstop\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    457\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    458\u001b[0m \u001b[43m        \u001b[49m\u001b[43miterator\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43miterator\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    459\u001b[0m \u001b[43m        \u001b[49m\u001b[43mchunksize\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunksize\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    460\u001b[0m \u001b[43m        \u001b[49m\u001b[43mauto_close\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mauto_close\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    461\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    462\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (\u001b[38;5;167;01mValueError\u001b[39;00m, \u001b[38;5;167;01mTypeError\u001b[39;00m, \u001b[38;5;167;01mLookupError\u001b[39;00m):\n\u001b[1;32m    463\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(path_or_buf, HDFStore):\n\u001b[1;32m    464\u001b[0m         \u001b[38;5;66;03m# if there is an error, close the store if we opened it.\u001b[39;00m\n",
      "File \u001b[0;32m~/.local/lib/python3.12/site-packages/pandas/io/pytables.py:906\u001b[0m, in \u001b[0;36mHDFStore.select\u001b[0;34m(self, key, where, start, stop, columns, iterator, chunksize, auto_close)\u001b[0m\n\u001b[1;32m    892\u001b[0m \u001b[38;5;66;03m# create the iterator\u001b[39;00m\n\u001b[1;32m    893\u001b[0m it \u001b[38;5;241m=\u001b[39m TableIterator(\n\u001b[1;32m    894\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m    895\u001b[0m     s,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    903\u001b[0m     auto_close\u001b[38;5;241m=\u001b[39mauto_close,\n\u001b[1;32m    904\u001b[0m )\n\u001b[0;32m--> 906\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mit\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_result\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.12/site-packages/pandas/io/pytables.py:2029\u001b[0m, in \u001b[0;36mTableIterator.get_result\u001b[0;34m(self, coordinates)\u001b[0m\n\u001b[1;32m   2026\u001b[0m     where \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mwhere\n\u001b[1;32m   2028\u001b[0m \u001b[38;5;66;03m# directly return the result\u001b[39;00m\n\u001b[0;32m-> 2029\u001b[0m results \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstart\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstop\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mwhere\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2030\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mclose()\n\u001b[1;32m   2031\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m results\n",
      "File \u001b[0;32m~/.local/lib/python3.12/site-packages/pandas/io/pytables.py:890\u001b[0m, in \u001b[0;36mHDFStore.select.<locals>.func\u001b[0;34m(_start, _stop, _where)\u001b[0m\n\u001b[1;32m    889\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mfunc\u001b[39m(_start, _stop, _where):\n\u001b[0;32m--> 890\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43ms\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mstart\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m_start\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstop\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m_stop\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mwhere\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m_where\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.12/site-packages/pandas/io/pytables.py:3291\u001b[0m, in \u001b[0;36mBlockManagerFixed.read\u001b[0;34m(self, where, columns, start, stop)\u001b[0m\n\u001b[1;32m   3288\u001b[0m dfs \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m   3290\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mnblocks):\n\u001b[0;32m-> 3291\u001b[0m     blk_items \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_index\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43mf\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mblock\u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[43mi\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43m_items\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3292\u001b[0m     values \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mread_array(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mblock\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mi\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m_values\u001b[39m\u001b[38;5;124m\"\u001b[39m, start\u001b[38;5;241m=\u001b[39m_start, stop\u001b[38;5;241m=\u001b[39m_stop)\n\u001b[1;32m   3294\u001b[0m     columns \u001b[38;5;241m=\u001b[39m items[items\u001b[38;5;241m.\u001b[39mget_indexer(blk_items)]\n",
      "File \u001b[0;32m~/.local/lib/python3.12/site-packages/pandas/io/pytables.py:2983\u001b[0m, in \u001b[0;36mGenericFixed.read_index\u001b[0;34m(self, key, start, stop)\u001b[0m\n\u001b[1;32m   2980\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mread_index\u001b[39m(\n\u001b[1;32m   2981\u001b[0m     \u001b[38;5;28mself\u001b[39m, key: \u001b[38;5;28mstr\u001b[39m, start: \u001b[38;5;28mint\u001b[39m \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m, stop: \u001b[38;5;28mint\u001b[39m \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   2982\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m Index:\n\u001b[0;32m-> 2983\u001b[0m     variety \u001b[38;5;241m=\u001b[39m _ensure_decoded(\u001b[38;5;28;43mgetattr\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mattrs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43mf\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[43mkey\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43m_variety\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m)\n\u001b[1;32m   2985\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m variety \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmulti\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m   2986\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mread_multi_index(key, start\u001b[38;5;241m=\u001b[39mstart, stop\u001b[38;5;241m=\u001b[39mstop)\n",
      "File \u001b[0;32m/usr/lib/python3/dist-packages/tables/attributeset.py:282\u001b[0m, in \u001b[0;36mAttributeSet.__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m    280\u001b[0m \u001b[38;5;66;03m# If attribute does not exist, raise AttributeError\u001b[39;00m\n\u001b[1;32m    281\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m name \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_v_attrnames:\n\u001b[0;32m--> 282\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mAttributeError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mAttribute \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mname\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[38;5;124m does not exist \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    283\u001b[0m                          \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124min node: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_v__nodepath\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    285\u001b[0m \u001b[38;5;66;03m# Read the attribute from disk. This is an optimization to read\u001b[39;00m\n\u001b[1;32m    286\u001b[0m \u001b[38;5;66;03m# quickly system attributes that are _string_ values, but it\u001b[39;00m\n\u001b[1;32m    287\u001b[0m \u001b[38;5;66;03m# takes care of other types as well as for example NROWS for\u001b[39;00m\n\u001b[1;32m    288\u001b[0m \u001b[38;5;66;03m# Tables and EXTDIM for EArrays\u001b[39;00m\n\u001b[1;32m    289\u001b[0m format_version \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_v__format_version\n",
      "\u001b[0;31mAttributeError\u001b[0m: Attribute 'block0_items_variety' does not exist in node: '/counts'"
     ]
    }
   ],
   "source": [
    "from cellphonedb.src.core.methods import cpdb_statistical_analysis_method\n",
    "from cellphonedb.src.core.methods import cpdb_analysis_method\n",
    "\n",
    "\n",
    "print(\"cellphone START\")\n",
    "print(datetime.datetime.now())\n",
    "\n",
    "cpdb_results = cpdb_statistical_analysis_method.call(\n",
    "    cpdb_file_path=\"../data/cellphonedb.zip\",  # use default database (or provide path to custom .zip)\n",
    "    meta_file_path=\"../output/cpdb/visium_meta.txt\",\n",
    "    counts_file_path=\"../output/cpdb/visium_counts.h5\",\n",
    "    counts_data=\"ensembl\",  # or 'ensembl' or gene_name\n",
    "    output_path=\"../output/cpdb/visium_cpdb\")\n",
    "\n",
    "\n",
    "print(\"cellphone END\")\n",
    "print(datetime.datetime.now())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
